### PR TITLE
Updated DCT implementation with documentation

### PR DESCRIPTION
**Pull request by rprechelt**
I updated the existing DCT implementation that someone contributed to be consistent with the rest of the project. I also added new docstrings to list parameter restrictions as defined by Accelerate.

### CODE REVIEW

**Comment by tkelman** [comment]
I think the plan terminology is used in Base with FFTW

**Comment by rprechelt** [comment]
I was looking closely at [FFTW.jl](https://github.com/JuliaLang/julia/blob/master/base/fft/FFTW.jl) while writing this, and yes, `plan_dct` is used by Base. 

I guess it comes down to whether this package is meant to be a drop-in replacement for Base functions or a true Julia interface to Accelerate that follows the calling conventions of Accelerate? The Accelerate version of `dct` supports extra forms/types then the one in Base and this should be to be exposed (but this could be done with optional arguments while still preserving `plan_dct` syntax).. 

I'm totally happy to rename this to be consistent with Base, but some clarification on the package standard for this issue might be helpful. @simonbyrne any thoughts?  There are a lot of small differences between Base and Accelerate with respect to the FFT, DFT and other functions, and having a standard to resolve this would be great.

**Comment by simonbyrne** [comment]
I'm not very familiar with FFTW, or the accelerate DSP functions, but if they are mostly used in a similar manner (even if there are some different optional arguments), then I think it would makes sense to use the same names.

**Comment by rprechelt** [comment]
Ok! I'll make the change momentarily and update the PR.

**Comment by rprechelt** [comment]
I updated the names to be consistent with Base.dct. I also added a destructor for DCT plan's that was not originally part of the commit. This has been incorporated into the testing.

File: src/DSP.jl [hunk 1]
```diff
@@ -1,5 +1,10 @@
 ## DSP.jl ##
 
+immutable DFT_Setup{T}
```
Comment (simonbyrne) [inline src/DSP.jl]: Julia convention typically avoids underscores where possible.

**Comment by rprechelt** [comment]
Changed DFT_Setup to DFTSetup to be more consistent with Julia conventions.

File: src/DSP.jl [hunk 2]
```diff
@@ -39,24 +44,63 @@ for (T, suff) in ((Float64, "D"), (Float32, ""))
...
+DCT 'length' must be equal to f*(2^n) where f = 1,3,5,15 and n >= 4. If you have a previous DCT setup
+object, that can be passed in as 'previous'. The returned DCT setup will share the underlying data
+storage of the previous setup object.
+
+Returns: DFTSetup
+"""
+function plan_dct(length::Integer,  dct_type::Integer, previous=C_NULL)
+    if dct_type < 2 &&  dct_type > 4
+        error("DCT type ", dct_type, " is not supported. Only DCT types 2, 3 and 4 are supported")
+    elseif !(isinteger(Base.log2(length)))
```
Comment (simonbyrne) [inline src/DSP.jl]: This check isn't correct. I think the easiest method would be:
```
n = trailing_zeros(length)
f = length >> n
if !(n >= 4 && f in (1,3,5,15))
...
```

Comment (rprechelt) [inline src/DSP.jl]: The check wasn't meant to be exhaustive, just a general check for some of the base conditions, but your method is better. I'll change it.

**Comment by rprechelt** [comment]
Changed.

File: src/DSP.jl [hunk 3]
```diff
@@ -39,24 +44,65 @@ for (T, suff) in ((Float64, "D"), (Float32, ""))
...
+    f = length >> n
+    if dct_type < 2 &&  dct_type > 4
+        error("DCT type ", dct_type, " is not supported. Only DCT types 2, 3 and 4 are supported")
+    elseif !(n >= 4 && f in (1,3,5,15))
+        error("Invalid DCT length. Length must be equal to f*(2^n) where f = 1,3,5,15 and n >= 4")
+    end
+    setup::Ptr{Void} = ccall(("vDSP_DCT_CreateSetup", libacc), Ptr{Void},
+                             (Ptr{Void}, UInt64, UInt64),
+                             previous, length, dct_type)
+    return DFTSetup{Float32}(setup, 0)
```
Comment (simonbyrne) [inline src/DSP.jl]: Standard practice is to attach a finalizer, so this gets cleaned up when the object is GC'ed. You may need to change the `DFTSetup` to a mutable type for this though.

Comment (simonbyrne) [inline src/DSP.jl]: Alternatively, if you would prefer to keep this more "low-level", you should document that the finalizer should be called manually.

Comment (rprechelt) [inline src/DSP.jl]: I like the finalizer method. Just implemented it now.

**Comment by simonbyrne** [comment]
Looks good to me. You'll need to rebase it though.

**Pull request merged**